src/drivers/intel/cavs/ssp.c

```diff
@@ -345,102 +346,57 @@ static inline int ssp_set_config(struct dai *dai,
 		goto out;
 	}
 
-	if (F_38400_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_38400_kHz / config->ssp.mclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 38.4MHz / %d Hz MCLK not divisable",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
+	/* MCLK config */
+	/* searching the smallest possible mclk source */
+	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
+		if (config->ssp.mclk_rate > ssp_freq[i].freq)
+			break;
 
-	if (F_38400_kHz % config->ssp.bclk_rate == 0) {
-		mdiv = F_38400_kHz / config->ssp.bclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 38.4MHz / %d Hz BCLK not divisable",
-				config->ssp.bclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-#elif defined(CONFIG_CANNONLAKE)
-	if (!config->ssp.mclk_rate || config->ssp.mclk_rate > F_24000_kHz) {
-		trace_ssp_error("ssp_set_config() error: "
-				"invalid MCLK = %d Hz (valid < 24000kHz)",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (!config->ssp.bclk_rate ||
-	    config->ssp.bclk_rate > config->ssp.mclk_rate) {
-		trace_ssp_error("ssp_set_config() error: "
-				"BCLK %d Hz = 0 or > MCLK %d Hz",
-				config->ssp.bclk_rate, config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
+		if (ssp_freq[i].freq % config->ssp.mclk_rate == 0) {
+			clk_index = i;
+			clk_set = 1;
```

Comment:
Could we limit the number of variables and initialize clk_index to -1 and check if it's >= 0?